### ISSUE TITLE
Validation mask in loss functions

### ISSUE THREAD

**Issue by ChrisSsak** [rpautrat/SuperPoint#303]
Hello, thank you very much for your amazing work!

I had a question regarding the validation mask in the detector and the descriptor loss:

![a](https://github.com/rpautrat/SuperPoint/assets/61319409/bdc351d6-0992-477e-be06-fa78b2008190)

Starting with the detector loss, I understand that the sum of the mask equates to Hc*Wc, which the detector loss is divided by.

What I do not understand is, from the picture, the summation is to Hc*Wc. 
But in the code, the loss is multiplied by the warped mask thus the borders are removed which is lower than Hc*Wc.

`valid_mask = tf.ones_like(keypoint_map) if valid_mask is None else valid_mask`

 I do not really understand why since the area of the mask is lower than Hc*Wc now. 

As for the descriptor loss, I understand the normalization part since the descriptor loss is divided by the H^2*W^2, but why multiply the loss by the mask?

Sorry if my wording is confusing and thank you again!

**Comment by rpautrat** [comment]
Hi,

For both the detector and descriptor losses, we only want to supervise pixels that are inside the valid mask. That is why we multiply the loss by the valid mask, effectively multiplying by 0 the pixels in the bordering artifacts. The normalization constant must then be adapted to the number of valid pixels, and not to Hc x Wc anymore.

The equations in the paper are not taking into account any bordering artifacts (it is actually unclear whether the original authors had bordering artifacts or not), so their formulation is simpler.

**Comment by ChrisSsak** [comment]
Thank you!

I will train two models with the warped validation mask and a validation mask of ones to see if there is significant impact. However, apologies for the second question, if I try a validation mask of ones, should I still do the same exact steps for the detector and descriptor losses?

Again, thank you for your reply!

**Comment by rpautrat** [comment]
If you don't use the valid mask, there are chances your network will have a poor training signal at the bordering artifacts (this is why we are masking them out in the loss).

If you still want to try that, then you can indeed keep the same steps as the current code. This will effectively divide by Hc x Wc in the normalization, so you would get the same as in the equations of the paper.

**Comment by ChrisSsak** [comment]
Thank you so much
